lib/fetch/util.js

```diff
@@ -297,6 +298,101 @@ function tryUpgradeRequestToAPotentiallyTrustworthyURL (request) {
   // TODO
 }
 
+function createDeferredPromise () {
+  let res
+  let rej
+  const promise = new Promise((resolve, reject) => {
+    res = resolve
+    rej = reject
+  })
+
+  return { promise, resolve: res, reject: rej }
+}
+
+// https://streams.spec.whatwg.org/#readablestreamdefaultreader-read-all-bytes
+/** @param {ReadableStreamDefaultReader} reader */
+async function ReadableStreamDefaultReaderReadAllBytes (reader) {
+  // 1. Let promise be a new promise.
+  const promise = createDeferredPromise()
+
+  // 2. Let bytes be an empty byte sequence.
+  const bytes = []
+
+  // 3. Read-loop given reader, bytes, and promise.
+  const readLoop = async () => {
+    const { value: chunk, done } = await reader.read()
+
+    // close steps - Resolve promise with bytes
+    if (done) {
+      promise.resolve(bytes)
+      return
+    }
+
+    // 1. If chunk is not a Uint8Array object, reject promise with a TypeError and abort these steps.
+    if (!isUint8Array(chunk)) {
+      // error steps, given e - Reject promise with e.
+      promise.reject(new TypeError('invalid chunk type received'))
+      return
+    }
+
+    // 2. Append the bytes represented by chunk to bytes.
+    bytes.push(chunk)
+
+    // 3. Read-loop given reader, bytes, and promise.
+    // This recursion could potentially cause a stack overflow if implemented directly. Implementations will need to mitigate this, 
+    // e.g. by using a non-recursive variant of this algorithm, or queuing a microtask, or using a more direct method of
+    // byte-reading as noted below. 
+    await new Promise(res => queueMicrotask(res))
+    return readLoop()
+  }
+
+  await readLoop()
+
+  return promise
+}
```

Comment:
From the look of it this function is going to be pretty inefficient as it allocates an enormous amount of promises.

Reply:
True, but I don't think in the normal case there should be any data to be read here at all.

Reply:
I agree, but unless we deviate from the spec, there isn't much else to do :(